## This notebook is to build the db for the crash data

In [1]:
#!/usr/bin/env python

import pandas as pd
from sodapy import Socrata
import matplotlib.pyplot as plt
from datetime import datetime
from modules.myfuncs import *

## Get the data from portal using Socrata client

In [2]:
# Unauthenticated client only works with public data sets. Note 'None'
# in place of application token, and no username or password:

url = "data.cityofchicago.org"
client = Socrata(url, None)

# Example authenticated client (needed for non-public datasets):
# client = Socrata(data.cityofchicago.org,
#                  MyAppToken,
#                  userame="user@example.com",
#                  password="AFakePassword")

## Grab preliminary data through query

In [3]:
# First 2000 results, returned as JSON from API / converted to Python list of
# dictionaries by sodapy.
# data website https://data.cityofchicago.org/resource/85ca-t3if.json

crash_data = client.get("85ca-t3if", 
                     #where='violation_date BETWEEN \'2015-01-01\' AND \'2020-12-01\'',
                     where="crash_date BETWEEN \'2015-01-01T00:00:00.000\' AND \'2020-12-31T00:00:00.000\'",
                     limit=1000000,
                    )

# Convert to pandas DataFrame
data = pd.DataFrame.from_records(crash_data)

## Investigate results

In [4]:
# 464k total
all_df = data.copy()

In [5]:
#client_df.dropna(subset=['camera_id']).isna().sum()
all_df.dropna(subset=['location',], inplace=True)
#all_df.isna().sum()

We dropped any crash that did not have lat/long location data.  It represents only 0.5% of data and would make it difficult to determine the intersection in which the crash occurred.

In [6]:
# drop a few columns we don't need, including location (we have lat/long)
dropme = ['statements_taken_i', 'private_property_i', 'photos_taken_i', 'dooring_i', 'date_police_notified','location']

all_df.drop(columns=dropme, inplace=True)

In [7]:
all_df.isna().sum()

crash_record_id                       0
rd_no                               774
crash_date                            0
posted_speed_limit                    0
traffic_control_device                0
device_condition                      0
weather_condition                     0
lighting_condition                    0
first_crash_type                      0
trafficway_type                       0
alignment                             0
roadway_surface_cond                  0
road_defect                           0
report_type                       11206
crash_type                            0
hit_and_run_i                    328412
damage                                0
prim_contributory_cause               0
sec_contributory_cause                0
street_no                             0
street_direction                      0
street_name                           0
beat_of_occurrence                    3
num_units                             0
most_severe_injury                  938


# What are the categories of 'traffic_control_device' and other columns
We care about 'traffic signal'.  When looking at crash data, these are at intersections.
We also will use intersection_related_i and use it to filter the crashes

In [8]:
# What's in this data?
col_interest = ['traffic_control_device', 'device_condition', 'weather_condition',
       'lighting_condition', 'first_crash_type', 'trafficway_type',
       'alignment', 'roadway_surface_cond', 'road_defect', 'report_type',
       'crash_type', 'hit_and_run_i', 'damage', 'prim_contributory_cause',
       'sec_contributory_cause', 'street_no', 'street_direction',
       'street_name', 'beat_of_occurrence', 'num_units', 'most_severe_injury', 
        'injuries_fatal', 'injuries_incapacitating',
       'injuries_non_incapacitating', 'injuries_reported_not_evident',
       'injuries_no_indication', 'injuries_unknown', 'crash_hour',
       'crash_day_of_week', 'crash_month', 'latitude', 'longitude', 'lane_cnt',
       'private_property_i', 'intersection_related_i', 'crash_date_est_i',
       'statements_taken_i', 'work_zone_i', 'work_zone_type',
       'workers_present_i']

for col in col_interest:
    print(col, all_df[col].unique())

traffic_control_device ['UNKNOWN' 'NO CONTROLS' 'TRAFFIC SIGNAL' 'STOP SIGN/FLASHER'
 'OTHER WARNING SIGN' 'LANE USE MARKING' 'OTHER' 'OTHER REG. SIGN'
 'SCHOOL ZONE' 'PEDESTRIAN CROSSING SIGN' 'YIELD'
 'FLASHING CONTROL SIGNAL' 'POLICE/FLAGMAN' 'RR CROSSING SIGN'
 'RAILROAD CROSSING GATE' 'OTHER RAILROAD CROSSING' 'DELINEATORS'
 'NO PASSING' 'BICYCLE CROSSING SIGN']
device_condition ['UNKNOWN' 'NO CONTROLS' 'FUNCTIONING PROPERLY' 'OTHER'
 'FUNCTIONING IMPROPERLY' 'NOT FUNCTIONING' 'WORN REFLECTIVE MATERIAL'
 'MISSING']
weather_condition ['UNKNOWN' 'CLEAR' 'SNOW' 'RAIN' 'CLOUDY/OVERCAST' 'FOG/SMOKE/HAZE'
 'OTHER' 'FREEZING RAIN/DRIZZLE' 'SLEET/HAIL' 'SEVERE CROSS WIND GATE'
 'BLOWING SNOW' 'BLOWING SAND, SOIL, DIRT']
lighting_condition ['DARKNESS, LIGHTED ROAD' 'DAYLIGHT' 'DARKNESS' 'DUSK' 'DAWN' 'UNKNOWN']
first_crash_type ['PARKED MOTOR VEHICLE' 'REAR END' 'ANGLE' 'SIDESWIPE SAME DIRECTION'
 'TURNING' 'REAR TO FRONT' 'PEDESTRIAN' 'PEDALCYCLIST' 'FIXED OBJECT'
 'SIDESWIPE OPPOSITE DIR

KeyError: 'private_property_i'

Let's drop the columns that we don't really care about.

In [ ]:
# remove na locations
print(len(all_df))
print(len(all_df.dropna(subset=['latitude'])))  # all have a location
#all_df.dropna(subset=['location'], inplace=True)  # drops about 2800 accidents with no recoreded location

intersection_related_i: A field observation by the police officer whether an intersection played a role in the crash. Does not represent whether or not the crash occurred within the intersection.

In [9]:
# How many were at traffic signals
signal_df = all_df[all_df['traffic_control_device']=='TRAFFIC SIGNAL']
other_df = all_df[all_df['traffic_control_device']!='TRAFFIC SIGNAL']
intrel_df = all_df[(all_df['intersection_related_i']=='Y')]
df = all_df[(all_df['intersection_related_i']=='Y')&(all_df['traffic_control_device']=='TRAFFIC SIGNAL')]

# how many crashes are at signals?
print('{:<25}{:<15}{:20}'.format('', 'n crash', 'Percent of total'))
print('{:<25}{:<15,}{:<20}'.format('All in study', len(all_df), '100%'))
print('{:<25}{:<15,}{:<4.1f}%'.format('At traffic signal', len(signal_df), len(signal_df)/len(all_df)*100))
print('{:<25}{:<15,}{:.1f}%'.format('Intersection-related', 
                                                    len(intrel_df), 
                                                    len(intrel_df)/len(all_df)*100
                                )
     )

print('{:<25}{:<15,}{:.1f}%'.format('Signal AND Intersection', 
                                                    len(df), 
                                                    len(df)/len(all_df)*100
                                )
     )
     


                         n crash        Percent of total    
All in study             463,798        100%                
At traffic signal        129,350        27.9%
Intersection-related     99,843         21.5%
Signal AND Intersection  60,332         13.0%


In [10]:
# what kind of crashes occur at traffic light, and how does it compare to other crashes?

def crash_stats(all_df, df):
    # by percentages
    crash_types = df['first_crash_type'].unique()

    print_list = []
    print('{:30}{:15}{:20}'.format("Type of Crash", 'Total crashes','Percent of total'))
    for crash in crash_types:
        num_crashes = len(df[df['first_crash_type']==crash])
        print_list.append([crash, num_crashes])
    
    print_list = sorted(print_list, key=lambda x: x[1], reverse=True)
    
    for crash, n in print_list: 
        print('{:30}{:<15d}{:>5.1f}%'.format(crash, n, 100 * n / len(df)))

def print_header(message):
    print('--'*20)
    print(message)
    print('--'*20)
        
def print_crash_stats(all_df, signal_df, other_df, signal_int_df):
    '''
    all_df Dataframe: all crashes in study 
    signal_df Dataframe: all signal related crashes
    other_df Dataframe: all non-signal related crahses
    signal_int_df: both signal and intersection related crashes
    '''
    print('CRASH STATS')
    print('Total crashes in study:', len(all_df))
    print('Signal crashes: {:.1f}%'.format((100 * len(signal_df) / len(all_df))))

    print_header('CRASHES NOT AT SIGNAL')
    print('Total crashes:', len(other_df))
    print()
    crash_stats(all_df, other_df)
    print('\n\n')
    
    
    print_header('CRASHES AT SIGNAL')
    print('Total crashes:', len(signal_df))
    print()
    crash_stats(all_df, signal_df)
    print('\n\n')

    
    print_header('CRASHES AT SIGNAL AND INTERSECTION RELATED')
    print('Total crashes:', len(signal_int_df))
    print()
    crash_stats(all_df, signal_int_df)
    

        

        
print_crash_stats(all_df, signal_df, other_df, df)


CRASH STATS
Total crashes in study: 463798
Signal crashes: 27.9%
----------------------------------------
CRASHES NOT AT SIGNAL
----------------------------------------
Total crashes: 334448

Type of Crash                 Total crashes  Percent of total    
PARKED MOTOR VEHICLE          103436          30.9%
REAR END                      61292           18.3%
SIDESWIPE SAME DIRECTION      50229           15.0%
ANGLE                         36606           10.9%
TURNING                       34481           10.3%
FIXED OBJECT                  18349            5.5%
PEDESTRIAN                    6639             2.0%
SIDESWIPE OPPOSITE DIRECTION  5725             1.7%
PEDALCYCLIST                  4854             1.5%
OTHER OBJECT                  3986             1.2%
HEAD ON                       3037             0.9%
REAR TO FRONT                 1995             0.6%
REAR TO SIDE                  1388             0.4%
OTHER NONCOLLISION            1352             0.4%
REAR TO REAR  

In [11]:
#!pip install geopy

In [12]:
from geopy.geocoders import Nominatim
geolocator = Nominatim(user_agent="https://github.com/sciencelee/chicago_rlc")


location = geolocator.geocode("175 5th Avenue NYC")
print(location.address)
# out: Flatiron Building, 175, 5th Avenue, Flatiron, New York, NYC, New York, ...

print((location.latitude, location.longitude))
# out: (40.7410861, -73.9896297241625)

#print(location.raw)
# out: {'place_id': '9167009604', 'type': 'attraction', ...}


# CAN USE THIS TO FIGURE OUT MY LAT LONG FROM RLC ADDRESS (or crash)


Flatiron Building, 175, 5th Avenue, Flatiron District, Manhattan Community Board 5, Manhattan, New York County, New York, 10010, United States
(40.741059199999995, -73.98964162240998)


In [13]:
# street no, street name, street direction?
for i in range(1):
    print(signal_df.iloc[i, :])
    
#     all_df['street_direction'].head() #N S E W
#     all_df['street_no'].head()  # just a number, so I guess street_direction goes with it.  Like 3700 N Ashland 
#     all_df['street_name'].head()

crash_record_id                  3358dc03b2fe52149c5901afd7221a9ebd509e3dfbaf49...
rd_no                                                                     JC366373
crash_date                                                 2019-07-27T02:16:00.000
posted_speed_limit                                                              30
traffic_control_device                                              TRAFFIC SIGNAL
device_condition                                              FUNCTIONING PROPERLY
weather_condition                                                            CLEAR
lighting_condition                                          DARKNESS, LIGHTED ROAD
first_crash_type                                                             ANGLE
trafficway_type                                                           FOUR WAY
alignment                                                       STRAIGHT AND LEVEL
roadway_surface_cond                                                           DRY
road

In [14]:
# BRINING SOME STUFF IN FROM THE RED CAMERA INVESTIGATION
# this loads up the lat/long of every camera
import pickle
import pandas as pd


with open('pickles/used_ints.pkl', 'rb') as f:
    used_ints = pickle.load(f)

with open('pickles/used_cams.pkl', 'rb') as f:
    used_cams = pickle.load(f)

#print(used_ints)
print(len(used_ints))
print(len(used_cams))

#print(used_ints)

183
363


In [15]:
# make interesections into a dataframe

ints_df = pd.DataFrame()

ints_df['intersection'] = used_ints.keys()
ints_df['location'] = ints_df['intersection'].apply(lambda x: used_ints[x][0])
ints_df['camera_ids'] = ints_df['intersection'].apply(lambda x: list(used_ints[x][1]))


ints_df['n_cams'] = ints_df['intersection'].apply(lambda x: len(used_ints[x][1]))
#cams_df['intersection'] = cams_df['address'].apply(lambda x: used_onts[x][2])


# we have way more cameras than intersections.  We need to consolidate
# some intersections have 3 or 4 cams.  Some have 1.
# we will focus on the intersection, not the individual camera for statistics
#print(len(ints_df.camera_id.unique()))
print(len(ints_df.intersection.unique()))

# NEED TO LOOK AT THE INTERSECTION NAMES TO MAKE SURE THEY MAKE SENSE
# WORRIED ABOUT NAMING CONVENTION 'ASHLAND AND CLARK' same as "CLARK AND ASHLAND"
ints_df.tail(50)
#print(ints_df.intersection.sort_values().unique())


# Data in this form works.  I have everything I need to start finding closest rlc to crash
# Take it back.  I do not have everything.  I really need to know the start and end dates for every intersection.
# This is tricky.  I need go through cam_ids, and find min start and max end for list.
# this would give me the start and end for each intersection.
# I am assuming, but do not know that cameras were installed as groups.  ASSUME??  Uh-oh. Need to know this.

# Rethinking this.  How about I just set up a database and go from there??

183


intersection                                  location  \
133           CANAL AND ROOSEVELT   (41.86717512472001, -87.63937581258092)   
134             119TH AND HALSTED  (41.677923389134385, -87.64198964584013)   
135              WESTERN AND 51ST  (41.801080203554356, -87.68381820888895)   
136         ROOSEVELT AND HALSTED   (41.86726309836901, -87.64697268601648)   
137          PULASKI AND MONTROSE   (41.96105580370759, -87.72789120598296)   
138        STONEY ISLAND AND 79TH   (41.75161152327319, -87.58559228428787)   
139        COTTAGE GROVE AND 71ST    (41.76580035131893, -87.6057291269901)   
140            HAMLIN AND MADISON   (41.88084147694413, -87.72090218457842)   
141    FULLERTON AND NARRAGANSETT   (41.92378618871128, -87.78556720562165)   
142           KEDZIE AND ARMITAGE   (41.91744617191702, -87.70711607187832)   
143        AUSTIN AND IRVING PARK   (41.95314405904539, -87.77692528423394)   
144              71ST AND ASHLAND   (41.76501830555763, -87.66385923012646)   
145              HALSTED AND 63RD   (41.77965750036339, -87.64493099278877)   
146          CICERO AND FULLERTON  (41.924347036010694, -87.74642763965628)   
147             WESTERN AND TOUHY  (42.012371334791766, -87.69039391737863)   
148              ASHLAND AND 47TH    (41.8085115972472, -87.66511729173287)   
149             ELSTON AND FOSTER  (41.975394608045924, -87.75174659886025)   
150           DAMEN AND FULLERTON   (41.92513494374925, -87.67796354553815)   
151               CICERO AND 47TH   (41.80772701677172, -87.74322936566276)   
152         CLARK AND IRVING PARK   (41.95448756687966, -87.66229349585691)   
153           PULASKI AND BELMONT   (41.939138153500274, -87.7273337101426)   
154          PULASKI AND LAWRENCE   (41.968262595100775, -87.7281753157596)   
155             OGDEN AND KOSTNER    (41.8480688449598, -87.73454745748458)   
156            ELSTON AND ADDISON   (41.946734754927796, -87.7086097735515)   
157              ASHLAND AND 95TH   (41.72116061205213, -87.66282813939165)   
158                DAMEN AND 63RD   (41.77922692069256, -87.67405273137854)   
159         LARAMIE AND FULLERTON   (41.92421579723144, -87.75621075458939)   
160              35TH AND WESTERN   (41.83018725526863, -87.68517792510231)   
161              PULASKI AND 63RD   (41.77863646878099, -87.72290635568541)   
162              ASHLAND AND 63RD   (41.77957043618697, -87.66429387512882)   
163            ADDISON AND HARLEM   (41.94533841246591, -87.80715476912044)   
164            PULASKI AND FOSTER  (41.975640616231914, -87.72833546926573)   
165              FOSTER AND NAGLE   (41.97568566971868, -87.78802040620296)   
166         CENTRAL AND FULLERTON   (41.92409019883488, -87.76605353016079)   
167              79TH AND HALSTED  (41.750549559438944, -87.64418836784098)   
168             CHICAGO AND CLARK  (41.896744346875806, -87.63132124815151)   
169           31ST AND CALIFORNIA  (41.837277050366296, -87.69517224651918)   
170          CENTRAL AND DIVERSEY   (41.93136555236339, -87.76639652879172)   
171           CENTRAL AND BELMONT   (41.93868377475747, -87.76656249384537)   
172          PULASKI AND ARMITAGE   (41.91727506940875, -87.72662274644799)   
173           CICERO AND ARMITAGE   (41.91697332451097, -87.74626022251057)   
174          DIVERSEY AND WESTERN   (41.93221698794738, -87.68803328089533)   
175              DAMEN AND ELSTON  (41.925731713685686, -87.67808889622799)   
176  NORTHWEST HIGHWAY AND FOSTER  (41.975153718866956, -87.76862036015007)   
177         CENTRAL AND MILWAUKEE  (41.976532211352456, -87.76862166095856)   
178         MILWAUKEE AND CENTRAL   (41.97535027477604, -87.76763298459159)   
179  FOSTER AND NORTHWEST HIGHWAY  (41.975791419680704, -87.76927689293382)   
180          MICHIGAN AND ONTARIO   (41.89342614670784, -87.62536202456768)   
181          MICHIGAN AND JACKSON   (41.87821734531058, -87.62445584029679)   
182         LAKE AND UPPER WACKER                 (41.8860672, -87.6365

In [16]:
# lets split the cameras and lat and long all up into separate columns so we can store it in a db
# we will just make 3 columns for cam1 cam2 cam3

ints_df['latitude'] = ints_df['location'].apply(lambda x: x[0])
ints_df['longitude'] = ints_df['location'].apply(lambda x: x[1])
ints_df['camera_id1'] = ints_df['camera_ids'].apply(lambda x: x[0])
ints_df['camera_id2'] = ints_df['camera_ids'].apply(lambda x: x[1] if len(x)>1 else None)
ints_df['camera_id3'] = ints_df['camera_ids'].apply(lambda x: x[2] if len(x)>2 else None)

In [17]:
ints_df
final_ints_df = ints_df.drop(columns=['camera_ids', 'location'])

In [18]:
final_ints_df

intersection  n_cams   latitude  longitude  \
0    31ST ST AND MARTIN LUTHER KING DRIVE       2  41.780231 -87.615640   
1                    ASHLAND AND DIVERSEY       2  41.932484 -87.668562   
2                 IRVING PARK AND LARAMIE       2  41.953866 -87.716609   
3                  ELSTON AND IRVING PARK       2  41.953838 -87.720427   
4              IRVING PARK AND KILPATRICK       2  41.953866 -87.716609   
..                                    ...     ...        ...        ...   
178                 MILWAUKEE AND CENTRAL       1  41.975350 -87.767633   
179          FOSTER AND NORTHWEST HIGHWAY       1  41.975791 -87.769277   
180                  MICHIGAN AND ONTARIO       2  41.893426 -87.625362   
181                  MICHIGAN AND JACKSON       2  41.878217 -87.624456   
182                 LAKE AND UPPER WACKER       2  41.886067 -87.636537   

    camera_id1 camera_id2 camera_id3  
0         2121       2123       None  
1         1623       1621       None  
2         1533       1534       None  
3         1503       1502       None  
4         2764       2763       None  
..         ...        ...        ...  
178       3022       None       None  
179       3003       None       None  
180       3084       3082       None  
181       3041       3043       None  
182       3051       3052       None  

[183 rows x 7 columns]

In [19]:
# Now put it into the db
# Create a db
conn = create_connection('database/rlc.db')  # function from myfuncs file
c = conn.cursor()
#conn.close()



sqlite3 version: 2.6.0
connected to database/rlc.db


## Create the table for intersection locations and cameras

In [20]:

def create_table(c, mytable, cols):
    '''
    Creates a new table 
    #Sets a constraint on db (UNIQUE) so that if you have camera_id and violation_date the same...
    '''
    
    flat_cols = []
    for col in cols:
        for item in col:
            flat_cols.append(item) 
    
    my_sql = 'CREATE TABLE {} ' + '(' + '{} ' * len(flat_cols) + ');'
    my_sql = my_sql.format(mytable, *flat_cols)  # insert datatypes for cols

    print(my_sql) # just to see what I'm doing

    
    try:     
        c.execute(my_sql) 
    except Exception as e:
        print('\nCREATE TABLE', mytable, 'FAILED!!',  e)


        
cols = [
        ['intersection', 'text'], 
        ['n_cams', 'int'],
        ['address', 'text'],
        ['latitude', 'real'],
        ['longitude', 'real'],
        ['camera_id1', 'int'],
        ['camera_id2', 'int'],
        ['camera_id3', 'int'],
        ]


create_table(c, 'intersection_cams', cols)
print()
conn.commit()

CREATE TABLE intersection_cams (intersection text n_cams int address text latitude real longitude real camera_id1 int camera_id2 int camera_id3 int );

CREATE TABLE intersection_cams FAILED!! table intersection_cams already exists



In [ ]:
sql_fetch_tables(c, conn)  # helper function in myfuncs
delete_all_entries(c, conn, 'intersection_cams')
final_ints_df.to_sql('intersection_cams', conn, if_exists='replace', index = False)


In [21]:
query = c.execute("SELECT * FROM intersection_cams;").fetchall()
print(query[:5])
print(len(query))

[('31ST ST AND MARTIN LUTHER KING DRIVE', 2, 41.7802309, -87.6156396, '2121', '2123', None), ('ASHLAND AND DIVERSEY', 2, 41.932484, -87.668562, '1623', '1621', None), ('IRVING PARK AND LARAMIE', 2, 41.9538663, -87.7166086, '1533', '1534', None), ('ELSTON AND IRVING PARK', 2, 41.953838, -87.720427, '1503', '1502', None), ('IRVING PARK AND KILPATRICK', 2, 41.9538663, -87.7166086, '2764', '2763', None)]
183


## Add table for all crashes

before finalizing my crash data, I want to tag all of the crashes that occured at my rlc intersections.

This might be difficult.


In [22]:
camloc_df = pd.read_sql_query("SELECT * from cam_locations", conn)


In [23]:
# what is the closest rlc intersection to each accident.  
# The code takes hours to run on macbook
    


# def closest_rlc(camloc_df, lat, long, m):
#     # m is meters distance threshold for determining at the correct intersection
#     latlong = (lat, long)
#     for i in range(len(camloc_df)):
#         loc = ((camloc_df.iloc[i]['latitude'], camloc_df.iloc[i]['longitude']))
#         if distance.distance(loc, latlong).meters < m:
#             return camloc_df.iloc[i]['intersection']
#             #print(loc, camloc_df.iloc[i]['intersection'], distance.distance(loc, latlong).meters)

#     else:
#         return None

    
    
# #print(crash_df.iloc[1, :])
# closest_rlc(camloc_df, df.iloc[1]['latitude'], df.iloc[1]['longitude'], 30)
# #df['intersection'] = 
# df[:5].apply(lambda x: closest_rlc(camloc_df, x.latitude, x.longitude, 30), axis=1)


In [24]:
# The code above works but is a memory HOG.  Let's try something else.
# maybe a simpler function would work??
# chi_lat = 41.8781
# deg_lat = 111070 # distance in meters
# deg_long = 83000 # distance in meters 

# def pythag_latlong(loc1, loc2):
#     delta_lat = abs(loc1[0] - loc2[0]) * deg_lat
#     delta_long = abs(loc1[1] - loc2[1]) * deg_long
#     return (delta_lat**2 + delta_long**2)**0.5

# def closest_rlc(camloc_df, lat, long, m):
#     # m is meters distance threshold for determining at the correct intersection
#     latlong = (lat, long)
#     for i in range(len(camloc_df)):
#         loc = ((camloc_df.iloc[i]['latitude'], camloc_df.iloc[i]['longitude']))
#         if pythag_latlong(loc, latlong) < m:
#             return camloc_df.iloc[i]['intersection']
#             #print(loc, camloc_df.iloc[i]['intersection'], distance.distance(loc, latlong).meters)

#     else:
#         return None


# intersect_list = []

# for i in range(10): #(len(df)):
#     intersect = closest_rlc(camloc_df, float(df.iloc[i]['latitude']), float(df.iloc[i]['longitude']), 30)
#     intersect_list.append(intersect)




Ended up using the code below due to memory problems.
Takes about 3 min to run on macbook

In [25]:
# Now I am desperate.  This takes too long to process.  Let's simplify it and make it a box instead.
box_side = 50  # effectively makes it check for crash being within 25m of interscection
box_lat = box_side / 111070 / 2 # 111070 is meters in deg lat in Chicago
box_long = box_side / 83000 / 2 # 83000 is meters in deg long in Chicago

def box_check(lat, long, ints_df):
    answer = (ints_df[  (ints_df['latitude'] > (lat - box_lat)) & 
                      (ints_df['latitude'] < (lat + box_lat)) &
                      (ints_df['longitude'] > (long - box_long)) &
                      (ints_df['longitude'] < (long + box_long))
                     ])
    if answer.empty: return None
    return answer['intersection'].values[0]
    
# THIS SEEMS TO WORK WITH SPEED AND ELIMINATES MEMORY PROBLEM
for i in range(5): #(len(df)):
    intersect = box_check(float(df.iloc[i]['latitude']), float(df.iloc[i]['longitude']), final_ints_df)
    print(intersect)
    
    
# MOMENT OF TRUTH
df['intersection'] = df.apply(lambda x: box_check(float(x.latitude), float(x.longitude), camloc_df), axis=1)



None
WESTERN AND 79TH
None
None
None


/Users/aaronlee/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [26]:
df['crash_date'] = pd.to_datetime(df['crash_date'])
df['year'] = df['crash_date'].apply(lambda x: int(x.year))
df['month'] = df['crash_date'].apply(lambda x: int(x.month))
df['day'] = df['crash_date'].apply(lambda x: int(x.day))
df['hour'] = df['crash_date'].apply(lambda x: int(x.hour))

/Users/aaronlee/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/Users/aaronlee/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/Users/aaronlee/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from 

In [27]:
sql_fetch_tables(c, conn)  # helper function in myfuncs


[('cam_startend',), ('cam_locations',), ('intersection_locations',), ('daily_violations',), ('hourly_congestion',), ('congestion_regions',), ('hourly_weather',), ('intersection_cams',), ('signal_crashes',)]


In [57]:
#camloc_df = pd.read_sql_query("SELECT * from cam_locations", conn)
regions_df  = pd.read_sql_query("SELECT * from congestion_regions", conn)

for col in ['west', 'east', 'south', 'north']:
    regions_df[col] = regions_df[col].apply(float)

df['latitude'] = df['latitude'].apply(float)
df['longitude'] = df['longitude'].apply(float)
   

regions_df.head()


/Users/aaronlee/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys
/Users/aaronlee/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


region region_id       west       east  \
0                      Lawndale N/S        14 -87.747456 -87.685372   
1   Washington Hts-Roseland-Pullman        26 -87.662508 -87.584845   
2                      Chicago Loop        13 -87.647208 -87.623080   
3  Beverly-Mt Greenwood-Morgan Park        25 -87.747456 -87.662508   
4    New City-Englewood-W Englewood        19 -87.683730 -87.636322   

       south      north                               description  
0  41.822792  41.866129   Pershing to Roosevel. Cicero to Western  
1  41.677477  41.728472    119th to 91st. Ashland to Stony Island  
2  41.866129  41.888860  Roosevelt to Wacker: Halsted to Michigan  
3  41.677477  41.728472          119th to 91st. Cicero to Ashland  
4  41.764066  41.822792      71st to Pershing. Western to Steward

In [96]:
df.columns
def which_region(lat, long, regions_df):
    #print(lat, long)
    row = regions_df[(regions_df['east'] >= long) &
                    (regions_df['west'] < long) &
                    (regions_df['north'] >= lat) &
                    (regions_df['south'] < lat)]['region_id'].max()
    return row

#df.iloc[:5]
# takes some 5min
df['region_id'] = df.apply(lambda x: which_region(x.latitude, x.longitude, regions_df), axis=1)

/Users/aaronlee/opt/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # This is added back by InteractiveShellApp.init_path()


In [97]:
df.head()

crash_record_id     rd_no  \
3   3358dc03b2fe52149c5901afd7221a9ebd509e3dfbaf49...  JC366373   
6   33597a24ed9ab58921e1f7516cc25b94720fd292923e1f...  JA513137   
15  335a69d78586c8d1d5d7f77a49b9643ded5f1889397ccc...  JD178846   
16  335a9396f8c88e17c504fa4065c240ecbb77f6cd86eb2a...  HZ376909   
19  0001301de281cb6fc82102d9f9decef9c0a40f5481c903...  JD252047   

            crash_date posted_speed_limit traffic_control_device  \
3  2019-07-27 02:16:00                 30         TRAFFIC SIGNAL   
6  2017-11-15 16:42:00                 35         TRAFFIC SIGNAL   
15 2020-03-08 20:31:00                 25         TRAFFIC SIGNAL   
16 2016-08-03 17:50:00                 30         TRAFFIC SIGNAL   
19 2020-06-03 08:00:00                 30         TRAFFIC SIGNAL   

        device_condition weather_condition      lighting_condition  \
3   FUNCTIONING PROPERLY             CLEAR  DARKNESS, LIGHTED ROAD   
6   FUNCTIONING PROPERLY             CLEAR                DARKNESS   
15  FUNCTIONING PROPERLY             CLEAR  DARKNESS, LIGHTED ROAD   
16  FUNCTIONING PROPERLY             CLEAR                DAYLIGHT   
19  FUNCTIONING PROPERLY              RAIN                DAYLIGHT   

            first_crash_type                  trafficway_type  ...  \
3                      ANGLE                         FOUR WAY  ...   
6   SIDESWIPE SAME DIRECTION  DIVIDED - W/MEDIAN (NOT RAISED)  ...   
15                   TURNING                      NOT DIVIDED  ...   
16  SIDESWIPE SAME DIRECTION                      NOT DIVIDED  ...   
19                   TURNING                         FOUR WAY  ...   

   crash_date_est_i work_zone_i work_zone_type workers_present_i  \
3               NaN         NaN            NaN               NaN   
6               NaN         NaN            NaN               NaN   
15              NaN         NaN            NaN               NaN   
16              NaN         NaN            NaN               NaN   
19              NaN         NaN            NaN               NaN   

        intersection  year month day hour region_id  
3               None  2019     7  27    2        14  
6   WESTERN AND 79TH  2017    11  15   16        22  
15              None  2020     3   8   20         3  
16              None  2016     8   3   17        12  
19              None  2020     6   3    8         3  

[5 rows x 49 columns]

In [98]:
cols = [
        ['crash_record_id', 'text'],
        ['rd_no', 'int'],
        ['crash_date', 'numeric'],
        ['posted_speed_limit', 'int'],
        ['traffic_control_device', 'text'],
        ['device_condition', 'text'],
        ['weather_condition', 'text'],
        ['lighting_condition', 'text'],
        ['first_crash_type', 'text'],
        ['trafficway_type', 'text'],
        ['alignment', 'text'],
        ['roadway_surface_cond', 'text'],
        ['road_defect', 'text'],
        ['report_type', 'text'],
        ['crash_type', 'text'],
        ['hit_and_run_i', 'text'],
        ['damage', 'text'],
        ['prim_contributory_cause', 'text'],
        ['sec_contributory_cause', 'text'],
        ['street_no', 'int'],
        ['street_direction', 'text'],
        ['street_name', 'text'],
        ['beat_of_occurrence', 'int'],
        ['num_units', 'int'],
        ['most_severe_injury', 'text'],
        ['injuries_total', 'int'],
        ['injuries_fatal', 'int'],
        ['injuries_incapacitating', 'int'],
        ['injuries_non_incapacitating', 'int'],
        ['injuries_reported_not_evident', 'int'],
        ['injuries_no_indication', 'int'],
        ['injuries_unknown', 'int'],
        ['crash_hour', 'int'],
        ['crash_day_of_week', 'int'],
        ['crash_month', 'int'],
        ['latitude', 'real'],
        ['longitude', 'real'],
        ['lane_cnt', 'int'],
        ['private_property_i', 'text'],
        ['intersection_related_i', 'text'],
        ['crash_date_est_i', 'text'],
        ['statements_taken_i', 'text'],
        ['work_zone_i', 'text'],
        ['work_zone_type', 'text'],
        ['workers_present_i', 'text'],
        ['intersection', 'text']  # THIS ONE STILL NEEDS TO BE BUILT
        ]


create_table(c, 'signal_crashes', cols)
print()
conn.commit()

CREATE TABLE signal_crashes (crash_record_id text rd_no int crash_date numeric posted_speed_limit int traffic_control_device text device_condition text weather_condition text lighting_condition text first_crash_type text trafficway_type text alignment text roadway_surface_cond text road_defect text report_type text crash_type text hit_and_run_i text damage text prim_contributory_cause text sec_contributory_cause text street_no int street_direction text street_name text beat_of_occurrence int num_units int most_severe_injury text injuries_total int injuries_fatal int injuries_incapacitating int injuries_non_incapacitating int injuries_reported_not_evident int injuries_no_indication int injuries_unknown int crash_hour int crash_day_of_week int crash_month int latitude real longitude real lane_cnt int private_property_i text intersection_related_i text crash_date_est_i text statements_taken_i text work_zone_i text work_zone_type text workers_present_i text intersection text );

CREATE TAB

In [99]:
sql_fetch_tables(c, conn)  # helper function in myfuncs
delete_all_entries(c, conn, 'signal_crashes')
df.to_sql('signal_crashes', conn, if_exists='replace', index = False)


[('cam_startend',), ('cam_locations',), ('intersection_locations',), ('daily_violations',), ('hourly_congestion',), ('congestion_regions',), ('hourly_weather',), ('intersection_cams',), ('signal_crashes',)]


In [ ]:
query = c.execute("SELECT * FROM signal_crashes;").fetchall()
for q in query[:2]:
    print(q)
    print()
print(len(query))

In [ ]:
cams_df[cams_df['intersection']=='VAN BUREN AND WESTERN']  
# FIXED mistake here!!!  I HAD two locations. 
# The 2nd location is wrong.  Latlongs were slightly shifted for multiple cams in same intersection

What I currently have:
- all crashes at signals with long/lat postions

What I want
- add a rlc to any crash that happened at traffic signal within x meters of a rlc.  I am assuming that anyting outside of 30m is at another signal.  This thinking makes sense for chicago streets which are typically 50 to 100 m E/W and about 150 to 200 m N/S


Can use with geopy to get distance


In [ ]:
# def closest_redlight(point, cams_df):
#     #distances = cams_df['location'].apply(lambda x: distance.distance(x, point).meters)
#     #return distances.min()  # this gives me the closest distance, but I want the closest camera

#     # 
#     distances = cams_df['camera_id'].apply(lambda x: [x, distance.distance(point, cams_df[cams_df['camera_id']==x]['location']).meters])
#     #closest = distances.apply(lambda x: x[1]).min()
#     #cam = distances[distances[0]==closet]
    
#     print(distances)
    


# signal_df2['latlong'][:1].apply(lambda x: closest_redlight(x, cams_df))

# ABOVE IS A CLOSE SOLUTON
# BELOW IS MY NEXT STEP.  I WOULD REALLY LIKE TO JUST RETURN THESE INDIVIDUALLY, BUT UNSURE IF I CAN.


# def close_redlight(point, cams_df):
#     #distances = cams_df['location'].apply(lambda x: distance.distance(x, point).meters)
#     #return distances.min()  # this gives me the closest distance, but I want the closest camera

#     # 
# #     distances = cams_df['camera_id'].apply(lambda x: [x, distance.distance(point, cams_df[cams_df['camera_id']==x]['location']).meters])
# #     my_min = min(distances, key=lambda x: x[1])
# #     #my_cam = distances[distances[0]==my_min]
    
#     # will see if loop is faster
#     for i in range(len(cams_df)):
#         print(cams_df.iloc[i]['location'])
    
#     return(my_min)
    

    
def closest_redlight(point, ints_df):

    # Try 3.  Will do a loop I can break from
    n = len(ints_df)
    threshold = 80
    
    for i in range(n):
        dist = distance.distance(ints_df.iloc[i]['location'], point).meters
        if dist < threshold:
            intersection = ints_df.iloc[i]['intersection'] 
            print(dist)
            break
    else:
        intersection = None 
      
    return intersection 
    
    

signal_df2['intersection'] = signal_df2['latlong'].apply(lambda x: closest_redlight(x, ints_df))

# TIMEIT ^^^^^^
# 100 rows takes 8s  (down from 1minute with original clunky code)
# We have 500k of them!!!!  That's 1000s or ~ 15min or so.

# TOMORROWS TASK
# add lat and long columns instead.  Don't measure distance p2p.  Lets just filter down to get ones less than threshold.
# we will end up checking a square basicaly, but it might be faster.


In [ ]:
print(signal_df2['intersection'].count(), len(signal_df2['intersection']))

In [ ]:
rlc_crash_df = signal_df2[signal_df2['intersection'].notnull()]
not_rlc_crash_df = signal_df2[signal_df2['intersection'].isnull()]
len(not_rlc_crash_df)

# Add in closest red light camera intersection to my crashes.
If a camera is within the threshold specified, it will recored the intersection.
This is key to being able to link my rlc data and my crash data.

In [ ]:
# what is the closest rlc intersection to each accident.  Is it under 50m
from geopy import distance



def closest_rlc(camloc_df, lat, long, m):
    # m is meters distance threshold for determining at the correct intersection
    latlong = (lat, long)
    for i in range(len(camloc_df)):
        loc = ((camloc_df.iloc[i]['latitude'], camloc_df.iloc[i]['longitude']))
        if distance.distance(loc, latlong).meters < m:
            return camloc_df.iloc[i]['intersection']
            #print(loc, camloc_df.iloc[i]['intersection'], distance.distance(loc, latlong).meters)

    else:
        return None

    
sql_fetch_tables(c, conn)  # helper function in myfuncs


camloc_df = pd.read_sql_query("SELECT * from cam_locations", conn)
    
#print(df.iloc[1, :])
closest_rlc(camloc_df, df.iloc[1]['latitude'], df.iloc[1]['longitude'], 50)
df['intersection'] = df.apply(lambda x: closest_rlc(camloc_df, x.latitude, x.longitude, 30), axis=1)

